samza-core/src/main/java/org/apache/samza/classloader/IsolatingClassLoaderFactory.java

```diff
@@ -121,31 +121,72 @@
   }
 
+  /**
+   * Build a classloader for loading {@link org.apache.samza.application.SamzaApplication} when in isolation mode.
+   *
+   * It is not sufficient to use the classloader from {@link #buildMainClassLoader()} to load the
+   * {@link org.apache.samza.application.SamzaApplication}, because the describe method might need to use classes in the
+   * framework infrastructure classloader (e.g. framework descriptors). The classloader from
+   * {@link #buildMainClassLoader()} would not be able to delegate from the application classloader to the framework
+   * infrastructure classloader.
+   *
+   * If the classloader from {@link #buildMainClassLoader()} is passed as the {@code parentClassLoader}, then the
+   * describe method can use framework infrastructure components and application-specific components.
+   *
+   * Implementation notes:
+   *
+   * This classloader will only directly load the {@link org.apache.samza.application.SamzaApplication} class, and it
+   * will delegate all other classloading to the {@code parentClassLoader}. This allows framework descriptors to be
+   * loaded from the framework infrastructure classloader. In addition, this allows the application processing classes
+   * to be loaded from the initial application classloader when calling describe. This is desirable because the core
+   * processing flows will delegate to the initial application classloader as well, and it is necessary that the same
+   * application classloader is used for all classes in the application processing flows.
+   *
+   * @param samzaApplicationClassName Class name of the {@link org.apache.samza.application.SamzaApplication}
+   * @param parentClassLoader {@link ClassLoader} to delegate to. Should be the {@link #buildMainClassLoader()}.
+   * @return {@link ClassLoader} to use for loading the {@link org.apache.samza.application.SamzaApplication}
+   */
+  public ClassLoader buildSamzaApplicationClassLoader(String samzaApplicationClassName,
+      ClassLoader parentClassLoader) {
+    File applicationLibDirectory = buildApplicationLibDirectory();
+    return LoaderBuilder.anIsolatingLoader()
+        // SamzaApplication is in the application classpath
+        .withClasspath(getClasspathAsURIs(applicationLibDirectory))
+        // getClasspathAsURIs should only return JARs within applicationLibDirectory anyways, but doing it to be safe
+        .withOriginRestriction(OriginRestriction.denyByDefault().allowingDirectory(applicationLibDirectory, false))
+        .withParentRelationship(DelegateRelationshipBuilder.builder()
+            .withDelegateClassLoader(parentClassLoader)
+            // almost everything should come from the parent classloader
+            .addDelegatePreferredClassPredicate(new GlobMatcher("*"))
+            // blacklisting the SamzaApplication class from the parent so that it gets loaded by this classloader
+            .addBlacklistedClassPredicate(new GlobMatcher(samzaApplicationClassName))
```

Comment:
Is it possible that by only blacklisting `samzaApplicationClass`, we end up class loading some of the dependencies of user code through parent class loader assuming both parent and application class loader are capable of classing loading that class?

Reply:
Ah yes, that is a good catch. I need to think through what the impact of that is.